### PR TITLE
Custom staging for GDS

### PR DESCRIPTION
**Pull request by mzient**
## Category:

**New feature** (*non-breaking change which adds functionality*)

**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
This PR enables custom staging engine for GDS.

The reader reads the numpy files to GDS registered chunks and then schedules a scatter-gather copy from the staging buffers to the user buffers. This should avoid host-side waits for copies between staging and target buffers, becasue these are now asynchronous, scheduled at a custom stream.

### Tests
- [X] Existing tests apply (Python)
- [X] New tests added (GTest)
  - [X] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [X] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2668

### CODE REVIEW

File: dali/test/python/nose_utils.py [hunk 1]
```diff
@@ -12,6 +12,18 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import nose
```
Comment (JanuszL) [inline dali/test/python/nose_utils.py:15]: Do we need this in this PR?

File: dali/operators/reader/gds_mem.cc [hunk 2]
```diff
@@ -12,40 +12,283 @@
...
 #include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/align.h"
 #include "dali/core/mm/malloc_resource.h"
 #include "dali/core/mm/composite_resource.h"
 
 namespace dali {
+namespace gds {
+
+static size_t GetGDSChunkSizeEnv() {
+  if (char *env = getenv("DALI_GDS_CHUNK_SIZE")) {
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:33]: I think we should document that in the op.

File: dali/operators/reader/gds_mem.cc [hunk 3]
```diff
@@ -12,40 +12,283 @@
...
 #include "dali/core/spinlock.h"
 #include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/align.h"
 #include "dali/core/mm/malloc_resource.h"
 #include "dali/core/mm/composite_resource.h"
 
 namespace dali {
+namespace gds {
+
+static size_t GetGDSChunkSizeEnv() {
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:32]: It can be moved to anonymous namespace. Why make it static?

File: dali/operators/reader/gds_mem.cc [hunk 4]
```diff
@@ -12,40 +12,283 @@
...
+    CUDA_CALL(cuFileBufDeregister(addr));
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// GDSRegisteredResource
+
+class GDSRegisteredResource : public mm::memory_resource<mm::memory_kind::device> {
+  void adjust_params(size_t &size, size_t &alignment) {
+    if (alignment < chunk_size_) {
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:70]: Is this condition sufficient? Maybe the alignment should be just a multiply of chunk_size_?

File: dali/operators/reader/gds_mem.cc [hunk 5]
```diff
@@ -12,40 +12,283 @@
...
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    char *raw = nullptr;
+    CUDA_CALL(cudaMalloc(&raw, size));
+    char *aligned = mm::detail::align_ptr(raw, alignment);
+    size_t padding = aligned - raw;
+    size_t chunks = (size - padding) / chunk_size_;
+    assert(chunks * chunk_size_ >= size);
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:85]: Protection against overflow?

File: dali/operators/reader/gds_mem.cc [hunk 6]
```diff
@@ -12,40 +12,283 @@
...
+  }
+
+  void do_deallocate(void *ptr, size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    std::unique_lock<spinlock> ul(map_lock_);
+    auto it = allocs_.find(ptr);
+    if (it == allocs_.end())
+      throw std::invalid_argument("The pointer was not allocated by this resource "
+                                  "or has been deleted.");
+    alloc_info alloc = it->second;
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:99]: Shouldn't we remove the element from `allocs_`?

File: dali/test/python/nose_utils.py (hunk 1 above)
Comment (mzient) [reply dali/test/python/nose_utils.py:15]: Not really, I can remove.

File: dali/operators/reader/gds_mem.cc (hunk 4 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:70]: `alignment` must be a power of 2 (and so does chunk_size), so if it's not less, then it's a multiple.

File: dali/operators/reader/gds_mem.cc (hunk 6 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:99]: We should, good spot.

File: dali/operators/reader/gds_mem.cc (hunk 5 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:85]: Rather checking that the padding didn't eat up too much.

File: dali/test/python/nose_utils.py (hunk 1 above)
Comment (JanuszL) [reply dali/test/python/nose_utils.py:15]: I would move it to a separate one if you want ot merge it.

File: dali/operators/reader/numpy_reader_gpu_op_impl.cu [hunk 7]
```diff
@@ -35,6 +35,9 @@ void NumpyReaderGPU::RunImplTyped(DeviceWorkspace &ws) {
   auto curr_batch = GetCurrBatchView<T, Dims>();
   const auto &dtype = TypeTable::GetTypeInfo<T>();
 
+
+  CUDA_CALL(cudaStreamWaitEvent(ws.stream(), staging_ready_));
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_gpu_op_impl.cu:39]: Understood.

File: dali/operators/reader/gds_mem.cc (hunk 3 above)
Comment (JanuszL) [reply dali/operators/reader/gds_mem.cc:32]: I thought that `static` is more C style, while namespace is more C++. Up to you.

File: dali/operators/reader/gds_mem.cc (hunk 5 above)
Comment (JanuszL) [reply dali/operators/reader/gds_mem.cc:85]: Understood.

File: dali/operators/reader/gds_mem.cc (hunk 4 above)
Comment (JanuszL) [reply dali/operators/reader/gds_mem.cc:70]: I would add a comment here about it. It is hard to infer this silent dependency between values without a deep understanding of the internals.

File: dali/operators/reader/gds_mem.cc [hunk 8]
```diff
@@ -12,40 +12,284 @@
...
+
+void GDSStagingEngine::return_unused(GDSStagingBuffer &&staging_buffer) {
+  std::lock_guard g(lock_);
+  ready_buffers_.insert(staging_buffer.release());
+}
+
+void GDSStagingEngine::copy_to_client(void *client_buffer, size_t nbytes,
+                                      GDSStagingBuffer &&staging_buffer, ptrdiff_t offset) {
+  std::lock_guard g(lock_);
+  assert(staging_buffer.at(0) != nullptr && "Staging buffer already released.");
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:233]: Should we check `staging_buffer.at(offset)` as well in this assert?

File: dali/operators/reader/gds_mem.h [hunk 9]
```diff
@@ -15,28 +15,156 @@
...
+ public:
+  explicit GDSStagingEngine(int device_id = -1, int max_buffers = 64, int commit_after = 32);
+  void set_stream(cudaStream_t stream);
+  GDSStagingBuffer get_staging_buffer(void *hint = nullptr);
+
+  /**
+   * @brief Enqueues a copy from a staging buffer to a client-provided destination buffer.
+   *
+   * Enqueues a copy from a staging buffer to a client buffer. The copy is enqueued, but not
+   * scheduled for execution until either a call to `commit` or a
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.h:121]: ``commit` or a...`?

File: dali/operators/reader/gds_mem_test.cc [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  size_t chunk = engn.chunk_size();
+  auto target_buf = mm::alloc_raw_unique<uint8_t, mm::memory_kind::device>(chunk * 2);
+  CUDA_CALL(cudaMemset(target_buf.get(), 0xcc, chunk * 2));
+  CUDA_CALL(cudaDeviceSynchronize());
+  engn.set_stream(stream);
+  auto buf0 = engn.get_staging_buffer();
+  auto buf1 = engn.get_staging_buffer();
+  engn.return_unused(std::move(buf0));
+  auto buf2 = engn.get_staging_buffer(buf1.at(chunk));
+  CUDA_CALL(cudaMemsetAsync(buf1.at(0), 0xaa, chunk, stream));
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem_test.cc:54]: Can we expect that `buf0==buf2`?

File: dali/operators/reader/loader/numpy_loader_gpu.h [hunk 11]
```diff
@@ -62,44 +72,15 @@ struct NumpyFileWrapperGPU {
...
-      cuFileBufDeregister(it->first);
-    }
-    reg_buff_.clear();
-  }
+  using CUFileLoader<NumpyFileWrapperGPU>::CUFileLoader;
 
   void PrepareEmpty(NumpyFileWrapperGPU& tensor) override;
   void ReadSample(NumpyFileWrapperGPU& tensor) override;
 
  protected:
```
Comment (JanuszL) [inline dali/operators/reader/loader/numpy_loader_gpu.h:83]: You can remove this now.

File: dali/operators/reader/gds_mem.cc (hunk 6 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:99]: Done.

File: dali/operators/reader/gds_mem.cc (hunk 4 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:70]: Alignments being powers of 2 is rather a common knowledge.

File: dali/operators/reader/gds_mem.cc [hunk 12]
```diff
@@ -12,40 +12,302 @@
...
+        DALI_FAIL(make_string(
+          "DALI_GDS_CHUNK_SIZE must be a number, optionally followed by 'k' or 'M', got: ",
+          env));
+      }
+    }
+    size_t s = atoll(env);
+    if (env[len-1] == 'k')
+      s <<= 10;
+    else if (env[len-1] == 'M')
+      s <<= 20;
```
Comment (JanuszL) [inline dali/operators/reader/gds_mem.cc:48]: Maybe getting size from the env variable can be an utility function? I guess we may need to use this to adjust allocator strategies in the long run. (just an idea)

File: dali/operators/reader/numpy_reader_op.cc [hunk 13]
```diff
@@ -117,6 +117,11 @@ This operator can be used in the following modes:
   The ``gpu`` backend requires cuFile/GDS support (418.x driver family or newer). which is
   shipped with the CUDA toolkit starting from CUDA 11.4. Please check the GDS documentation
   for more details.
+
+  The ``gpu`` reader reads the files in chunks. The size of the chunk can be controlled
+  process-wide with an environment variable ``DALI_GDS_CHUNK_SIZE``. Valid values are powers of 2
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_op.cc:122]: What are the tradeoff when increasing the chunk size? Maybe you can add an about it too?

File: dali/operators/reader/gds_mem.cc (hunk 12 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:48]: Perhaps, but maybe in another PR?

Comment (JanuszL) [reply dali/operators/reader/gds_mem.cc:48]: Sure, just an idea.

File: dali/operators/reader/gds_mem.cc (hunk 2 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc]: Done.

File: dali/operators/reader/numpy_reader_op.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/reader/numpy_reader_op.cc:122]: I don't have any high quality info about that, just some general idea - nothing that would justify inflating the documentation. When we have some data (and with GDS it's not easy to come by) we can add it.

[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/numpy_reader_op.cc:122]: Understood.

**Review by szalpal — APPROVED** [review]

File: dali/test/python/nose_utils.py (hunk 1 above)
Comment (mzient) [reply dali/test/python/nose_utils.py]: Done.

File: dali/operators/reader/gds_mem.cc (hunk 4 above)
Comment (mzient) [reply dali/operators/reader/gds_mem.cc:86]: Done.

File: dali/operators/reader/loader/numpy_loader_gpu.h (hunk 11 above)
Comment (mzient) [reply dali/operators/reader/loader/numpy_loader_gpu.h]: Done.

**Pull request merged**